### ISSUE TITLE
(aws-batch): Most JobDefinition.ContainerProperties.LinuxParameters are not rendered

### ISSUE THREAD

**Issue by damonmaria** [aws/aws-cdk#13023]
`JobDefinition.buildJobContainer()` explicitly only renders `devices` from `container.linuxParams`, omitting all other parameters.

The implementation for `JobDefinition` makes it seem like this is intentional. 

### Reproduction Steps

```typescript
new JobDefinition(this, 'jobDef', {
    container: {
        linuxParams: new LinuxParameters(this, 'linuxParams', {
          sharedMemorySize: 24_000,
        }),
        ...
    },
    jobDefinitionName: 'test',
})
```  

### What did you expect to happen?

The rendered stack should contain:
```json
          "LinuxParameters": {
            "SharedMemorySize": 24000
          },
```

### What actually happened?

No parameters are rendered:
```json
          "LinuxParameters": {},
```

### Environment

  - **CDK CLI Version: 1.89.0**
  - **Framework Version: 1.89.0**
  - **Node.js Version: v12.20.2** 
  - **OS               : MacOS 11.2.1**
  - **Language (Version): Typescript 4.1.5** 

### Other

Here's the relevant code in `job-definition.js`:
```javascript
    buildJobContainer(container) {
        if (container === undefined) {
            return undefined;
        }
        return {
            ...
            linuxParameters: container.linuxParams
                ? { devices: container.linuxParams.renderLinuxParameters().devices }
                : undefined,
            ...
```

--- 

This is :bug: Bug Report

**Comment by damonmaria** [comment]
Just FYI I extended `JobDefinition` and overrode `buildJobContainer` changing the `linuxParameters` to:
```javascript
      linuxParameters: container.linuxParams
        ? container.linuxParams.renderLinuxParameters()
        : undefined,
```

And it worked no problem.

**Comment by iliapolo** [comment]
@damonmaria Would you be up to whip up a PR with that fix? :)

**Comment by damonmaria** [comment]
Given that someone spent the time to only pull devices (pulling everything would have been much easier) I'm assuming it was done like that for a reason. There's no comment and I'm not sure if that reason is still relevant. 

That made me a bit weary of doing a PR.

**Comment by iliapolo** [comment]
Yeah I understand that. I can't think of a reason it would intentionally be this way, and it feels like an oversight to me. 

We can ping the original author though to be sure.

@stephnr any thoughts on this?

**Comment by stephnr** [comment]
Hi! I will take a look at this soon ✌️

**Comment by stephnr** [comment]
So - I can see that at the time of creation, this was built to support only devices and no other linux parameters. However, it should be an easy fix to have this extended. I will give it a shot 👌 

https://docs.aws.amazon.com/batch/latest/APIReference/API_LinuxParameters.html

**Comment by stephnr** [comment]
Yep, so I can confirm that I will try switching over to the ECS construct for Linux Params. 

However, they are missing some newer params that are now supported (i.e. Swappiness). So I'll look into adding these along with the change.

**Comment by iliapolo** [comment]
@stephnr sounds good. But can I ask you to add the new parameters to ECS in a separate PR? We usually stick to a single change per PR (helps with reverts and changelog).

**Comment by stephnr** [comment]
Sure, no problem 😊👌

**Comment by stephnr** [comment]
Hm - I am having issues with getting this supported. Seems like there is a diff between the generated Batch Job Definition spec of LinuxParams vs. the ECS definition. See below:

```
$ cdk-build
[2021-02-25T23:30:14.394] [ERROR] jsii/compiler - Compilation errors prevented the JSII assembly from being created
lib/job-definition.ts:440:7 - error TS2322: Type 'LinuxParametersProperty | undefined' is not assignable to type 'IResolvable | LinuxParametersProperty | undefined'.
  Type 'LinuxParametersProperty' is not assignable to type 'IResolvable | LinuxParametersProperty | undefined'.
    Type 'import("/Users/Stephen/Documents/github/aws-cdk/packages/@aws-cdk/aws-ecs/lib/ecs.generated").CfnTaskDefinition.LinuxParametersProperty' is not assignable to type 'import("/Users/Stephen/Documents/github/aws-cdk/packages/@aws-cdk/aws-batch/lib/batch.generated").CfnJobDefinition.LinuxParametersProperty'.
      Types of property 'tmpfs' are incompatible.
        Type 'import("/Users/Stephen/Documents/github/aws-cdk/packages/@aws-cdk/core/lib/resolvable").IResolvable | (import("/Users/Stephen/Documents/github/aws-cdk/packages/@aws-cdk/core/lib/resolvable").IResolvable | import("/Users/Stephen/Documents/github/aws-cdk/packages/@aws-cdk/aws-ecs/lib/ecs.generated").CfnTaskDefinition....' is not assignable to type 'import("/Users/Stephen/Documents/github/aws-cdk/packages/@aws-cdk/core/lib/resolvable").IResolvable | (import("/Users/Stephen/Documents/github/aws-cdk/packages/@aws-cdk/core/lib/resolvable").IResolvable | import("/Users/Stephen/Documents/github/aws-cdk/packages/@aws-cdk/aws-batch/lib/batch.generated").CfnJobDefiniti...'.
          Type '(IResolvable | TmpfsProperty)[]' is not assignable to type 'IResolvable | (IResolvable | TmpfsProperty)[] | undefined'.
            Type '(import("/Users/Stephen/Documents/github/aws-cdk/packages/@aws-cdk/core/lib/resolvable").IResolvable | import("/Users/Stephen/Documents/github/aws-cdk/packages/@aws-cdk/aws-ecs/lib/ecs.generated").CfnTaskDefinition.TmpfsProperty)[]' is not assignable to type '(import("/Users/Stephen/Documents/github/aws-cdk/packages/@aws-cdk/core/lib/resolvable").IResolvable | import("/Users/Stephen/Documents/github/aws-cdk/packages/@aws-cdk/aws-batch/lib/batch.generated").CfnJobDefinition.TmpfsProperty)[]'.
              Type 'import("/Users/Stephen/Documents/github/aws-cdk/packages/@aws-cdk/core/lib/resolvable").IResolvable | import("/Users/Stephen/Documents/github/aws-cdk/packages/@aws-cdk/aws-ecs/lib/ecs.generated").CfnTaskDefinition.TmpfsProperty' is not assignable to type 'import("/Users/Stephen/Documents/github/aws-cdk/packages/@aws-cdk/core/lib/resolvable").IResolvable | import("/Users/Stephen/Documents/github/aws-cdk/packages/@aws-cdk/aws-batch/lib/batch.generated").CfnJobDefinition.TmpfsProperty'.
                Type 'TmpfsProperty' is not assignable to type 'IResolvable | TmpfsProperty'.
                  Type 'import("/Users/Stephen/Documents/github/aws-cdk/packages/@aws-cdk/aws-ecs/lib/ecs.generated").CfnTaskDefinition.TmpfsProperty' is not assignable to type 'import("/Users/Stephen/Documents/github/aws-cdk/packages/@aws-cdk/aws-batch/lib/batch.generated").CfnJobDefinition.TmpfsProperty'.
                    Types of property 'containerPath' are incompatible.
                      Type 'string | undefined' is not assignable to type 'string'.
                        Type 'undefined' is not assignable to type 'string'.

440       linuxParameters: container.linuxParams
          ~~~~~~~~~~~~~~~
Error: /Users/Stephen/Documents/github/aws-cdk/node_modules/jsii/bin/jsii --silence-warnings=reserved-word exited with error code 1
Build failed.!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error: last command failed. fix problem and resume by executing: /Users/Stephen/Documents/github/aws-cdk/scripts/foreach.sh
directory:    /Users/Stephen/Documents/github/aws-cdk/packages/@aws-cdk/aws-batch
```

Primarily, this part:

```
Types of property 'containerPath' are incompatible.
    Type 'string | undefined' is not assignable to type 'string'.
        Type 'undefined' is not assignable to type 'string'.
```

@iliapolo - how do you advise I moved forward with this...? 🤔

**Comment by stephnr** [comment]
This is blocked until we can rebuild the generated definition file for Batch LinuxParameters.

**Comment by iliapolo** [comment]
> This is blocked until we can rebuild the generated definition file for Batch LinuxParameters.

What do you mean? our spec seems up to date. 

The problem is that the [`containerPath`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-taskdefinition-tmpfs.html#cfn-ecs-taskdefinition-tmpfs-containerpath) property from ECS is optional, while the [`containerPath`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-batch-jobdefinition-tmpfs.html#cfn-batch-jobdefinition-tmpfs-containerpath) property from Batch is required - so they cannot be assigned to one another. 

This makes me think we might not have gone the right road of re-using ECS interfaces. Is there an inherent reason to make this coupling? if its just for the sake of not redeclaring the same interfaces, we might need to reconsider...

**Comment by stephnr** [comment]
Yeah - the initial thought here was to keep both in sync along the same
interface for LinuxParameters. However, it seems we cannot go that route.
Perhaps we go the route of defining a custom interface for the batch
package?

- Stephen

**Comment by dbrezack** [comment]
It looks like this is also occurring for the addCapabilities option as well
```
    const myLinuxParams = new ecs.LinuxParameters(this, 'myLinuxParams');
    myLinuxParams.addCapabilities(ecs.Capability.MKNOD, ecs.Capability.SYS_ADMIN);
    myLinuxParams.addDevices({
      hostPath: '/dev/fuse',
      permissions: [ecs.DevicePermission.READ, ecs.DevicePermission.WRITE,ecs.DevicePermission.MKNOD]
    });
```
Devices works perfectly but addCapabilities does not. Also as per the documentation our launch type is EC2 https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-ecs.LinuxParameters.html#addwbrcapabilitiescap

```
          "LinuxParameters": {
            "Devices": [{"HostPath": "/dev/fuse", "Permissions": ["read", "write","mknod"]}]
          },
```

**Comment by iliapolo** [comment]
@stephnr 

> Perhaps we go the route of defining a custom interface for the batch
package?

Yeah, I think we have to break this dependency. It seems to only exist for the sake of reusing interfaces, but doesn't exemplify a proper integration between the two services. 

We need to remove all references to ECS interfaces. This is actually critical work before we can make this package stable. Good thing we discovered this. You have capacity to take this on?

**Comment by stephnr** [comment]
Yep - makes sense. I have capacity now to work on this. Its definitely a medium/high effort but I have made a lot of headway. I will try committing some work now so we can get a review process started on the methodology. @iliapolo

**Comment by stephnr** [comment]
@iliapolo an early look into some of the changes can be seen via the referenced commit above. Overall, my plan is to redefine some of the resources we use from ECS to within the Batch package. But, in an effort to keep it simple, I am removing some of the redundancies found within the ECS package (i.e. [Abstract interfaces on AWS SecretsManager](https://github.com/aws/aws-cdk/blob/84c9960b56b98d365b45399e1f230355c1e87014/packages/%40aws-cdk/aws-ecs/lib/container-definition.ts#L20))

**Comment by iliapolo** [comment]
@stephnr 

> I am removing some of the redundancies found within the ECS package

We cannot remove any public API from this package as it is already stable. I'd also advice against that regardless, let's focus on one thing at a time.

Thanks for your engagement here!

**Comment by stephnr** [comment]
Sure. Let's take this one piece at a time. I think the first area to target is the container properties for a JobDefinition.

There is definitely a lot of differences here vs. ECS. How would you advise we tackle this against the public API?

**Comment by iliapolo** [comment]
@stephnr 

> How would you advise we tackle this against the public API?

Not sure I follow. The batch package is not stable yet, we can freely remove public API's. (I was referring to ECS as the stable package we cannot break anymore). 

Or did you mean something else?

**Comment by stephnr** [comment]
Ah, no I meant only altering the batch package. I have no plans to alter anything in any of the other packages.

Sorry if that wasn't clear 😅

**Comment by iliapolo** [comment]
Phew!

Anyway, sounds good :)

**Comment by stephnr** [comment]
@iliapolo I made a ton of headway today and just finished pushing up my changes. They are ready for an early review (if you have time). It still needs to be updated for our tests but this is roughly 90% of the way it's heading right now.

Next up was to implement a few soft checks on resource creation to make sure the user doesn't place values out of bounds or such.

**Comment by iliapolo** [comment]
@stephnr Thanks! 

I'll try to get to it promptly 👍

**Comment by damonmaria** [comment]
Giving this a bump. Any advancement? By chance does aws-cdk v2 solve this?

**Comment by stephnr** [comment]
Still waiting on expert review from AWS. - Bump @iliapolo ❤️

**Comment by iliapolo** [comment]
@stephnr You mind creating a PR for this change? I don't mind its incomplete or even doesn't compile, just makes it easier for me to look at and checkout locally. 

Thanks

**Comment by dbrezack** [comment]
Any updates on this issue ?

**Comment by stephnr** [comment]
This is on my court to finish up. I haven't had much spare time but I need to setup a PR with the adjustments.

**Comment by yashwanthkp** [comment]
Any updates on this will be great.

**Comment by stephnr** [comment]
Sorry for the delay here. Lately my schedule hasn't had much room to continue on this.

I'm happy to let someone else address this.

**Comment by suzhoum** [comment]
Hi bumping this issue! Is there an update? We need to increase the size of shared memory, but at the moment, we have to manually edit the template and use CF to deploy, otherwise it's not rendered or deployed correctly with `cdk synth` or `cdk deploy`. Is there a better workaround?

**Comment by f0cus10** [comment]
Bumping this issue. 

In the meantime looks like we have to implement our own custom L2 construct since `buildJobContainer()` is a private function and cannot be overridden. 

Any simpler workarounds?

**Comment by jamie1911** [comment]
Hello, any updates on this? It is a bit disheartening that we are still in experimental stages for batch JobDefinition properties to be fully supported. Is it going to be supported in cdk v3?

**Comment by comcalvi** [comment]
this has been fixed in the version that will be released this week

**Comment by damonmaria** [comment]
Confirm this is correctly output now
